doc/acl/Extend-L3V6ACLs.md

```diff
@@ -0,0 +1,308 @@
+     * | MATCH_ICMPV6_CODE    |      √       |
+     * | MATCH_ICMP_TYPE      |      √       |
+     * | MATCH_ICMP_CODE      |      √       |
+     * |-------------------------------------|
+     * | MATCH_IP_PROTOCOL    |      √       |
+     * | MATCH_NEXT_HEADER    |      √       |
+     * | ------------------------------------|
+     * | MATCH_L4_SRC_PORT    |      √       |
+     * | MATCH_L4_DST_PORT    |      √       |
+     * | MATCH_TCP_FLAGS      |      √       |
+     * |-------------------------------------|
+     */
+
+
+##### Pros and Cons
+All the pros and cons of option-B (extending L3V6 ACL table) applies here as well.
+Additionally, option-C has the below cons:
+- Need platform checks to determine which platforms can support combined v4 and v6 ACL table.
+- Even on platforms that support combined v4 and v6 ACL tables, we need additional checks to identify which platforms are optimized to have v4 and v6 in the same ACL table.
+
+#### Implementation
+
+Based on the above design considerations, option-B is implemented.
+Additionally, a new field is added to the ACL capability in STATE_DB to help applications identify the platforms where v4 fields can be matched in L3V6 ACL tables without additional hardware costs.
+
+##### Orchagent
+
+Today, during the initialization of AclOrch, the default built-in ACL table types are created. In this proposal, using platform specific checks, AclOrch identifies platforms where v4 fields can be matched in L3V6 ACL tables without additional hardware costs. On these identified platforms v4 matchfields are supported in L3V6 table.
+
+  ``` aclorch::init() -> initDefaultTableTypes() -> addAclTableType(TABLE_TYPE_L3V6) {```
+  
+  ```
+   
+            :
+            :
+            if ( /* platform optimizes v4 in L3V6Table */) {
+              .withMatch(make_shared<AclTableMatch>(SAI_ACL_TABLE_ATTR_FIELD_ETHER_TYPE))
+              .withMatch(make_shared<AclTableMatch>(SAI_ACL_TABLE_ATTR_FIELD_SRC_IP))
+              .withMatch(make_shared<AclTableMatch>(SAI_ACL_TABLE_ATTR_FIELD_DST_IP))
+              .withMatch(make_shared<AclTableMatch>(SAI_ACL_TABLE_ATTR_FIELD_ICMP_TYPE))
+              .withMatch(make_shared<AclTableMatch>(SAI_ACL_TABLE_ATTR_FIELD_ICMP_CODE))
+            }
+            :
+            :
+    }
+  
+   ```
+
+
+##### STATE_DB
+
+A new field called `optimized_V4_in_L3V6` is added to the ACL capability in STATE_DB. This field is set to true by orchagent during AclOrch init on platforms where v4 fields can be matched in L3V6 ACL tables without additional hardware costs. This provides an interface to the operator to identify the platforms where the operator can choose to add IPv4 ACL rules in L3V6 ACL tables. In the future, this field can be used to validate configuration workflows to prevent a user from adding L3 ACL rules to L3V6 ACL table on unsupported platforms.
+
+```
+127.0.0.1:6379[6]> hgetall "ACL_STAGE_CAPABILITY_TABLE|INGRESS"
```

Comment:
How can orchagent know if the combined L3V4V6 can be supported? Are we planning to add a SAI API for querying the capability?

Reply:
SAI community does not have a consensus to support an API for this specific capability check.
Please refer https://github.com/opencomputeproject/SAI/pull/1408#issue-1126526787.

So we plan to add platform specific checks.